internal/drivers/azure/driver.go

```diff
@@ -0,0 +1,344 @@
+import (
+	"context"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/drone-runners/drone-runner-aws/internal/drivers"
+	"github.com/drone-runners/drone-runner-aws/internal/lehelper"
+	"github.com/drone-runners/drone-runner-aws/internal/oshelp"
+	"github.com/drone-runners/drone-runner-aws/types"
+	"github.com/drone/runner-go/logger"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/to"
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/compute/armcompute"
+)
+
+type config struct {
+	init sync.Once
+
+	tenantID          string
+	clientID          string
+	clientSecret      string
+	subscriptionID    string
+	resourceGroupName string
+
+	rootDir string
+
+	location string // region, example: East US
+
+	// image data
+	publisher string
+	offer     string
+	sku       string
+	version   string
+
+	IPAddress string
+
+	size        string
+	tags        map[string]string
+	zones       []*string
+	userData    string
+	userDataKey string
+
+	username string
+	password string
+
+	service *armcompute.VirtualMachinesClient
+	cred    azcore.TokenCredential
+}
+
+func New(opts ...Option) (drivers.Driver, error) {
+	p := new(config)
+	for _, opt := range opts {
+		opt(p)
+	}
+
+	if p.service == nil {
+		cred, err := azidentity.NewClientSecretCredential(p.tenantID, p.clientID, p.clientSecret, nil)
+		p.cred = cred
+		if err != nil {
+			return nil, err
+		}
+
+		p.service, err = armcompute.NewVirtualMachinesClient(p.subscriptionID, cred, nil)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return p, nil
+}
+
+func (c *config) RootDir() string {
+	return c.rootDir
+}
+
+func (c *config) DriverName() string {
+	return string(types.Azure)
+}
+
+func (c *config) InstanceType() string {
+	return c.offer
+}
+
+func (c *config) CanHibernate() bool {
+	return false
+}
+
+func (c *config) Zones() string {
+	var z string
+	if len(z) == 1 {
+		return *c.zones[0]
+	}
+	for _, zone := range c.zones {
+		z = fmt.Sprintf("%s%s,", z, *zone)
+	}
+
+	return z
+}
+
+func (c *config) Create(ctx context.Context, opts *types.InstanceCreateOpts) (instance *types.Instance, err error) {
+	var name = fmt.Sprintf(opts.RunnerName+"-"+opts.PoolName+"-%d", time.Now().Unix())
+
+	vnetName := fmt.Sprintf("%s-vnet", name)
+	subnetName := fmt.Sprintf("%s-subnet", name)
+	publicIPName := fmt.Sprintf("%s-publicip", name)
+	networkInterfaceName := fmt.Sprintf("%s-networkinterface", name)
+	diskName := fmt.Sprintf("%s-disk", name)
+
+	var tags = map[string]*string{}
+	// add user defined tags
+	var t string
+	for k, v := range c.tags {
+		t = v
+		tags[k] = &t
+	}
+
+	logr := logger.FromContext(ctx).
+		WithField("cloud", types.Azure).
+		WithField("name", name).
+		WithField("image", c.InstanceType()).
+		WithField("pool", opts.PoolName).
+		WithField("zone", "").
+		WithField("image", c.offer).
+		WithField("size", c.size)
+
+	logr.Debugln("Starting Azure Setup")
```

Comment:
Maybe better to have this in info mode rather than debug (if it's not printed anywhere else?)

Reply:
done